content/browser/attribution_reporting/aggregatable_attribution_utils.cc

```diff
@@ -121,6 +129,114 @@ std::vector<AggregatableHistogramContribution> CreateAggregatableHistogram(
   return contributions;
 }
 
+
+// TODO(kelly): Organize in classes
+void AttributionLogicLastTouch(Partition& partition, 
+        base::flat_map<std::string, std::vector<absl::uint128>>& trigger_keypieces_per_source) {
+
+  // for (uint64_t i=attribution_window.epoch_start(); i<=attribution_window.epoch_end(); i++) {
+  //   for (auto& pair : epoch_sources_count[i]) {
+  //     auto& source_key_piece = pair.first;
+  //     auto count = pair.second;
+
+  //     auto it = source_counts.find(source_key_piece);
+  //     if (it == source_counts.end()) {
+  //       source_counts[source_key_piece] = count;
+  //     }
+  //     source_counts[source_key_piece] += count;
+  //   }
+  // }
+}
+
+void AttributionLogicUniform(Partition& partition,
+        base::flat_map<std::string, std::vector<absl::uint128>>& trigger_keypieces_per_source) {
+
+  auto& attribution_window = partition.attribution_window;
+  auto& sources_per_epoch = partition.sources_per_epoch;
+  
+  double total_sources_count = 0;
+
+  base::flat_map<std::string, base::flat_map<absl::uint128, double>>
+          source_counts_per_sourcekey;
+
+
+  for (uint64_t i=attribution_window.epoch_start(); 
+          i <= attribution_window.epoch_end(); i++) {
+    
+    // Ignore empty epochs
+    auto it = sources_per_epoch.find(i);
+    if (it == sources_per_epoch.end()) {
+      continue;
+    }
+
+    // Count occurrences per source keypiece across all epochs
+    for (StoredSource* source : sources_per_epoch[i]) {
+      auto aggregation_keys = source->aggregation_keys().keys();
+      for (auto& pair : aggregation_keys) {
+        auto& source_key = pair.first;
+        auto& key_piece = pair.second;
+
+        auto it1 = source_counts_per_sourcekey.find(source_key);
+        if (it1 == source_counts_per_sourcekey.end()) {
+          source_counts_per_sourcekey[source_key] = {};
+        }
+
+        auto& source_counts = source_counts_per_sourcekey[source_key];
+        auto it2 = source_counts.find(key_piece);
+        if (it2 == source_counts.end()) {
+          source_counts[key_piece] = 0;
+        }
+        source_counts[key_piece] += 1;
+      }
```

Comment:
How do `find()` and `end()` work here, is it some C++ syntax or some ARA methods for these objects?

Reply:
it's c++ STL stuff